src/turbomind/models/llama/BlockManager.cc

```diff
@@ -86,7 +86,9 @@ size_t BlockManager::GetBlockCount(size_t block_size, double ratio)
     size_t free{};
     size_t total{};
     check_cuda_error(cudaMemGetInfo(&free, &total));
-    return static_cast<size_t>(total * ratio) / block_size;
+    FT_CHECK_WITH_INFO(total * ratio - (total - free) > 0,
```

Comment:
I agree. But should it be `total * ratio <= free`?

Reply:
There's no need to have this restriction, it doesn't matter.
e.g. A100 80G, Llama 2 13b Chat, ratio 0.9
`total * ratio` =80*0.9=72, `free` =80-26=54, `total * ratio - (total - free)` =46 (weights: 26g, kv cache: 46g)
In this case, `total * ratio` is not less than `free` but it's ok.
We may take `total * ratio - (total - free)` as a whole.

Reply:
Maybe we can consider revising the definition of `ratio`. Using the ratio*total-sized memory for k/v cache can impose a significant mental burden on users. For example, when running Llama 2 7b chat fp16 on an A30 24g device, the weights occupy 14g, and the default ratio parameter is set to 0.5. If not configured, it could result in OOM. Even when configured manually, the user would need to subtract the weights from the total memory, make estimations, and perform calculations to determine an appropriate `ratio`. It's not user-friendly.
The `ratio` can convey the following meanings, eliminating the need for users to consider model weights.
```
weights: total - free
k/v cache: total * ratio - (total - free)
other: (1 - ratio) * total
```

Reply:
Maybe the `ratio` definition could be the size of `ratio*total - MODEL_WEIGHTS` memory is allocated for the k/v cache.